File: src/accessTokens.ts
import jwt from 'jsonwebtoken';
import {TOKENS_URL} from './config';
import {RefreshToken} from './interfaces';
import {makeRequest} from './makeRequest';

let accessTokenCache: {[key: string]: string} = {};

export function resetAccessTokenCache(): void {
  accessTokenCache = {};
}

export async function getAccessToken(
  refreshToken: RefreshToken,
  platform: string,
  scope: string,
): Promise<string> {
  const cacheKey = `${ platform }:${ scope }`;

  let accessToken: string = accessTokenCache[cacheKey];

  if (!accessToken || hasExpired(accessToken)) {
    accessToken = await createAccessToken(refreshToken, scope, platform);
    accessTokenCache[cacheKey] = accessToken;
  }

  return accessToken;
}

async function createAccessToken(
  refreshToken: RefreshToken,
  platform: string,
  scope: string,
): Promise<string> {
  const {token: accessToken} = await makeRequest<{token: string}>({
    url: TOKENS_URL,
    method: 'POST',
    token: refreshToken.token,
    body: {
      subject: refreshToken.subject,
      platform,
      scope,
    },
  });

  return accessToken;
}

function hasExpired(token: string): boolean {
  const {exp: expiration} = (jwt.decode(token) || {}) as {exp?: number};

  const now = Math.floor(Date.now() / 1000);

  if (expiration && now - expiration >= 0) {
    return true;
  }

  return false;
}

File: src_py/grndwork_api_client/access_tokens.py
import time
from typing import Dict

import jwt

from .config import TOKENS_URL
from .dtos import RefreshToken
from .make_request import make_request


access_token_cache: Dict[str, str] = {}


def reset_access_token_cache() -> None:
    global access_token_cache
    access_token_cache = {}


def get_access_token(
    refresh_token: RefreshToken,
    platform: str,
    scope: str,
) -> str:
    cache_key = f'{platform}:{scope}'
    access_token = access_token_cache.get(cache_key)

    if not access_token or has_expired(access_token):
        access_token = create_access_token(refresh_token, platform, scope)
        access_token_cache[cache_key] = access_token

    return access_token


def create_access_token(
    refresh_token: RefreshToken,
    platform: str,
    scope: str,
) -> str:
    result, _ = make_request(
        url=TOKENS_URL,
        method='POST',
        token=refresh_token['token'],
        body={
            'subject': refresh_token['subject'],
            'platform': platform,
            'scope': scope,
        },
    )

    access_token = result.get('token')

    if isinstance(access_token, str):
        return access_token

    raise ValueError('No token returned')


def has_expired(token: str) -> bool:
    decoded_token = jwt.decode(
        token,
        algorithms=['HS256'],
        options={'verify_signature': False},
    )

    expiration = int(decoded_token.get('exp', 0))
    now = int(time.time() * 1000)

    if expiration and now - expiration >= 0:
        return True

    return False
